autometa/taxonomy/ncbi.py

```diff
@@ -366,47 +372,64 @@ def is_common_ancestor(self, taxid_A, taxid_B):
-        f"{args.query_taxid} Lineage:\n{ncbi.lineage(taxid=args.query_taxid)}\n"
-        f"{args.query_taxid} is_common_ancestor {query_taxid_parent}: "
-        f"{ncbi.is_common_ancestor(parent_taxid=query_taxid_parent, child_taxid=args.query_taxid)}\n"
-    )
+        lineage_a_taxids = {ancestor.get("taxid") for ancestor in self.lineage(taxid_A)}
+        lineage_b_taxids = {ancestor.get("taxid") for ancestor in self.lineage(taxid_B)}
+        common_ancestor = lineage_b_taxids.intersection(lineage_a_taxids)
+        common_ancestor.discard(1)  # This discards root
+        return True if common_ancestor else False
+
+    def is_valid_taxid(self, taxid):
+        """
+        Checks if the given `taxid` is a positive integer or not as well as if it is present in nodes.dmp or names.dmp
+        In case the taxid provided is in the form of a string the function will cast the string as an integer
+
+        Parameters
+        ----------
+        taxid : int
+            identifer for a taxon in NCBI taxonomy databases - nodes.dmp, names.dmp or merged.dmp
+
+        Returns
+        -------
+        int
+            `taxid` if the `taxid` is a positive integer and present in either nodes.dmp or names.dmp or
+            new_taxid which is equal to the secondary `taxid` present in merged.dmp
+
+        Raises
+        ------
+        ValueError
+            Provided `taxid` is not a positive integer
+        """
+        # This checks if an integer has been added as str, eg. "562"
+        if isinstance(taxid, str):
+            invalid_chars = set(string.punctuation + string.ascii_letters)
+            invalid_chars.discard(".")
+            if {char for char in taxid if char in invalid_chars} or taxid.count(
+                "."
+            ) > 1:
+                raise ValueError(f"taxid contains invalid character(s)! Given: {taxid}")
+            taxid = float(taxid)
+        # a boolean check is needed as something can be a boolean but not necessarily a str, eg. TRUE/FALSE
+        # float(taxid).is_integer() checks if it is something like 12.0 vs. 12.9
+        # is_integer only takes float as input else raises error, thus isinstance( ,float) is used before it to make sure a float is being passed
+        if isinstance(taxid, bool) or (
+            isinstance(taxid, float) and not taxid.is_integer()
+        ):
+            raise ValueError(f"taxid must be a positive integer! Given: {taxid}")
+        taxid = int(taxid)
+        if taxid <= 0:
+            raise ValueError(f"Taxid must be a positive integer! Given: {taxid}")
+        if not self.names.get(taxid) and not self.nodes.get(taxid):
+            new_taxid = self.merged.get(taxid, taxid)
+            if not self.names.get(new_taxid) and not self.nodes.get(new_taxid):
+                logger.warning(
+                    f"Taxid : {taxid} not found in either nodes.dmp or names.dmp !"
+                )
+            return int(new_taxid)
```

Comment:
I don't like this - if the taxid is not found in both `nodes.dmp` and `merged.dmp`, then there is a warning but the function still returns the original taxid given to it. So the client functions will not know if this is a valid taxid. I think the `is_valid_taxid` function should return `None` if the taxid is not valid, and then in the client functions there should be error handling for this eventuality.

Reply:
I think I am getting a bit lost. I understand that a `taxid` may be old or fused with something else and its newer version may be present in `merged.dmp` and that's why we search that and return the latest one in case the old one is absent from `nodes.dmp` or `names.dmp`. 
As per the current implementation of `is_valid_taxid`, even after searching through `merged.dmp` if the `taxid` is not found in the database then the old taxid is returned which is handled by the client function. 

The `get` command (example attached) in each of the functions using `is_valid_taxid` handles the case when `taxid` is not present in their respective database. This is done by returning, `1` in case of `parent` function, `unclassified` in `rank` function, and so on.

Example of `rank` function using `get`
https://github.com/KwanLab/Autometa/blob/aaa7ff5c435be39fdaf95fe7c401e6f8a1789ee8/autometa/taxonomy/ncbi.py#L235

From what I understand is that by logging the error the user will have the option to know why some `taxids` are not being classified. Given the `get` command in each of the functions using `is_valid_taxid` handles the case when `taxid` is not present in the database I am not sure why returning `taxid` is a bad idea.